tools/osbuilder/rootfs-builder/rootfs.sh

```diff
@@ -713,6 +717,17 @@ EOF
 	info "Create /etc/resolv.conf file in rootfs if not exist"
 	touch "$dns_file"
 
+	if [ "${GUEST_PULL}" == "yes" ]; then
```

Comment:
@ChengyuZhu6 did I understand correctly from https://github.com/kata-containers/kata-containers/issues/8104 and https://github.com/kata-containers/kata-containers/issues/8103 that we always need the pause image in the rooftf due to security, performance and a limitation on containerd? If yes than this guard should be removed, right?

I really appreciate if you could add an in-code comment with the rationale to include the pause image in the rootfs; I think such as important decision should be recorded alongside the doc :)

@fitzthum could you please review this piece as you raised the security concern?

Reply:
I think we need this check to avoid putting the pause image in the rootfs for the non-CoCo cases.

For CoCo it seems like we are planning to always use the pause image from the rootfs. This is important because the pause image executes inside the pod context. I'm not sure that the rootfs script is the most significant piece. The rootfs is measured so we don't have to worry too much about whether the pause image is there (from a security standpoint). To me the bigger question is how we can be sure that the pause image that we run always come from the rootfs. I think this has more to do with the behavior of the agent/image-rs. I haven't looked through this PR yet, but I will try to double-check the behavior there.

Reply:
Yes, I agree with @fitzthum. For non-CoCo cases, the pause image in the rootfs is an extra feature that is not required.

>I really appreciate if you could add an in-code comment with the rationale to include the pause image in the rootfs; I think such as important decision should be recorded alongside the doc :)

Thanks for your suggestion. I will do.

Reply:
>To me the bigger question is how we can be sure that the pause image that we run always come from the rootfs. I think this has more to do with the behavior of the agent/image-rs. I haven't looked through this PR yet, but I will try to double-check the behavior there.

@fitzthum By enabling the GUEST_PULL option, we can instruct the kata-runtime to obtain the guest pull image information from nydus-snapshotter, as shown below:
```
{
  {
  "volume_type": "image_guest_pull",
  "options": [
    "containerd.io/snapshot/cri.layer-digest=sha256:24fb2886d6f6c5d16481dd7608b47e78a8e92a13d6e64d87d57cb16d5f766d63",
    "containerd.io/snapshot/nydus-proxy-mode=true"
  ],
  "image_pull": {
    "metadata": {
      "containerd.io/snapshot/cri.layer-digest": "sha256:24fb2886d6f6c5d16481dd7608b47e78a8e92a13d6e64d87d57cb16d5f766d63",
      "containerd.io/snapshot/nydus-proxy-mode": "true"
         }
       }
  }
}
```

Next, the kata-runtime will send the packaged information to the kata-agent, as illustrated below:
```
    "driver": "image_guest_pull", 
    "driver_options": [
        "image_guest_pull"{
            "metadata":{
                "containerd.io/snapshot/cri.layer-digest": "sha256:24fb2886d6f6c5d16481dd7608b47e78a8e92a13d6e64d87d57cb16d5f766d63",
                "containerd.io/snapshot/nydus-proxy-mode": "true",
                "io.katacontainers.pkg.oci.bundle_path": "/run/containerd/io.containerd.runtime.v2.task/k8s.io/cb0b47276ea66ee9f44cc53afa94d7980b57a52c3f306f68cb034e58d9fbd3c6",
                "io.katacontainers.pkg.oci.container_type": "pod_container",
                "io.kubernetes.cri.container-name": "coco-container",
                "io.kubernetes.cri.container-type": "container",
                "io.kubernetes.cri.image-name": "quay.io/kata-containers/confidential-containers:unsigned",
                "io.kubernetes.cri.sandbox-id":"7a0d058477e280604ae02de6a016959e8a05fcd3165c47af41eabcf205b55517",
                "io.kubernetes.cri.sandbox-name": "coco-pod","io.kubernetes.cri.sandbox-namespace": "default",
                "io.kubernetes.cri.sandbox-uid": "de7c6a0c-79c0-44dc-a099-69bb39f180af",
            }
        }
    ], 
    "source": "quay.io/kata-containers/confidential-containers:unsigned", 
    "fstype": "overlay", 
    "options": [], 
    "mount_point": "/run/kata-containers/cb0b47276ea66ee9f44cc53afa94d7980b57a52c3f306f68cb034e58d9fbd3c6/rootfs", 
```

After that, the kata-agent will handle the image_guest_pull volume and invoke the imageService to pull the image.
```
let image_service = image::ImageService::singleton().await?;
let bundle_path = image_service
            .pull_image(image_name, &cid, &image_pull_volume.metadata)
            .await?;
```

The `pull_image` function in `image.rs` would check the container type. If the container is for sandbox, the image service will use the pause image from the rootfs.
```
let is_sandbox = if let Some(value) = image_metadata.get("io.kubernetes.cri.container-type")
        {
            value == "sandbox"
        } else if let Some(value) = image_metadata.get("io.kubernetes.cri-o.ContainerType") {
            value == "sandbox"
        } else {
            false
        };

        if is_sandbox {
            let mount_path = Self::unpack_pause_image(cid, "pause")?;
            self.add_image(String::from(image), String::from(cid)).await;
            return Ok(mount_path);
        }
```

The behavior ensures that the pause image that we run always come from the rootfs.

Reply:
@ChengyuZhu6 I really liked the sequence of operations and sample code you pasted here, this help us a lot to understand the implementation and review this PR. I was thinking in put this all in a doc to help other reviewer as well as future maintenance. What do you think?

Back to my original question, now I realize it doesn't make sense to remove the `GUEST_PULL` guard because of non-CoCo use case at least. Fine. However I still have some doubts regarding that guard and the pause image. Here is...

... the pull image by host sharing approach which first implementation you've introduced in https://github.com/kata-containers/kata-containers/pull/7837 ... does it need the pause image embedded in the rootfs too? Same question for kata-agent and image-rs linking, does the host sharing leverage the guest-components/image-rs ?

What I'm trying to figure out is how to build Kata for the cases of enabling:

a) guest pull only
b) host sharing only
c) both guest pull and host sharing

For example, whether for the host sharing case we will need to export `GUEST_PULL=yes` too  so that pause image is pulled in rootfs at build time and kata-agent is linked with image-rs; this way we won't be able to build the case b) alone.

Reply:
>I was thinking in put this all in a doc to help other reviewer as well as future maintenance. What do you think?
@wainersm That sounds like a great idea and I would be grateful for your help in doing this.

Reply:
@wainersm Thanks for your questions.

>the pull image by host sharing approach which first implementation you've introduced in https://github.com/kata-containers/kata-containers/pull/7837 ... does it need the pause image embedded in the rootfs too?

For the first question, the host sharing approach doesn't need the pause image in the rootfs.

>Same question for kata-agent and image-rs linking, does the host sharing leverage the guest-components/image-rs ?

Regarding the second question, the host sharing approach does not require exporting `GUEST_PULL=yes`. The host sharing approach has two features: sharing with dm-verity check and sharing without dm-verity check. In the CCv0 branch, the host sharing without dm-verity check can work on both CoCo and non-CoCo scenarios in Kata. However, the host sharing with dm-verity check depends on the image-rs. For merging the feature to the main branch, I am working on moving the dm-verity feature from image-rs to Kata (https://github.com/confidential-containers/guest-components/issues/394). This means that we will not rely on image-rs at all when the work is done.

Reply:
Even in the case of host sharing, we should still get the pause image from inside the initrd, right? In theory we could use dm-verity hashes to verify the pause image, but doesn't seem convenient for the user. If we want the pause image in the rootfs for host pulling, is this conditional the right one?

Reply:
that's my concern @fitzthum , that if host sharing uses the pause image pre-pulled inside the image then the conditional (GUEST_PULL) is conceptually wrong. Not a big deal though.